crates/bpf-filtering/src/process_tree.rs

```diff
@@ -24,16 +32,55 @@ pub(crate) enum Error {
     ProcessNotFound { pid: Pid },
     #[error("loading process {pid}: parent image {ppid} not found")]
     ParentNotFound { pid: Pid, ppid: Pid },
+    #[error(transparent)]
+    Procfs(#[from] ProcfsError),
+    #[error("failed to get the {ns_type} namespace for process {pid}")]
+    Namespace { pid: Pid, ns_type: String },
 }
 
 pub(crate) const PID_0: Pid = Pid::from_raw(0);
 
+fn get_process_namespace(pid: Pid, ns_type: &str) -> Result<u32, Error> {
+    let path = Path::new("/proc")
+        .join(pid.to_string())
+        .join("ns")
+        .join(ns_type);
+
+    let link_target = fs::read_link(path).map_err(|_| Error::Namespace {
+        pid,
+        ns_type: ns_type.to_owned(),
+    })?;
+    let link_target = link_target.to_string_lossy();
+    let ns: u32 = NAMESPACE_RE
+        .captures(&link_target)
+        .and_then(|cap| cap.get(1))
+        .and_then(|m| m.as_str().parse().ok())
+        .ok_or(Error::Namespace {
+            pid,
+            ns_type: ns_type.to_owned(),
+        })?;
+
+    Ok(ns)
+}
+
+fn get_process_namespaces(pid: Pid) -> Result<Namespaces, Error> {
+    Ok(Namespaces {
+        uts: get_process_namespace(pid, "uts").unwrap_or_default(),
```

Comment:
For now I'm doing `unwrap_or_default` here, because apparently some processes have malformed namespace data, like:

```shell
$ sudo -E ./target/release/pulsar-exec pulsard
[2023-09-25T01:06:12Z INFO  pulsar::pulsard::daemon] Starting module process-monitor
[2023-09-25T01:06:12Z ERROR pulsar::pulsard::module_manager] Error in module process-monitor. Module stopped. Error initializing process filtering

    Caused by:
        failed to get the uts namespace for process 11358

pulsar on  namespaces [$!] via  v1.72.0 took 2s
❯ sudo ls -la /proc/11358/ns/
ls: cannot read symbolic link '/proc/11358/ns/net': No such file or directory
ls: cannot read symbolic link '/proc/11358/ns/uts': No such file or directory
ls: cannot read symbolic link '/proc/11358/ns/ipc': No such file or directory
ls: cannot read symbolic link '/proc/11358/ns/pid_for_children': No such file or directory
ls: cannot read symbolic link '/proc/11358/ns/mnt': No such file or directory
ls: cannot read symbolic link '/proc/11358/ns/cgroup': No such file or directory
ls: cannot read symbolic link '/proc/11358/ns/time': No such file or directory
ls: cannot read symbolic link '/proc/11358/ns/time_for_children': No such file or directory
total 0
dr-x--x--x 2 root       root       0 Sep 25 03:04 .
dr-xr-xr-x 9 vadorovsky vadorovsky 0 Sep 25 02:48 ..
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 cgroup
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 ipc
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 mnt
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 net
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 pid -> 'pid:[4026532723]'
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 pid_for_children
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 time
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 time_for_children
lrwxrwxrwx 1 root       root       0 Sep 25 03:07 user -> 'user:[4026532721]'
lrwxrwxrwx 1 root       root       0 Sep 25 03:04 uts

pulsar on  namespaces [$!] via  v1.72.0
❯ sudo cat /proc/11358/comm
zypak-sandbox

pulsar on  namespaces [$!] via  v1.72.0
❯ pstree -s -p 11358
systemd(1)───systemd(3137)───gnome-shell(3287)───bwrap(11301)───bwrap(11322)───brave(11323)───zypak-sandbox(11358)
```

I'm wondering if we should log a warning in such case. The thing is that we don't seem to have logs at all in the bpf-filtering crate, but I could add it if you are fine with that @banditopazzo

Reply:
Alright, I will use `log::warn` when.